### PR TITLE
Add container size check in ThriftBinaryProtocol

### PR DESCRIPTION
**Pull request by rubanm**
The read length check is currently enabled in ThriftBinaryDeserializer.

https://github.com/twitter/elephant-bird/blob/master/core/src/main/java/com/twitter/elephantbird/thrift/ThriftBinaryDeserializer.java#L46

However, we do not check the container sizes against this. As a result, invalid container sizes returned by the underlying Transport can cause OutOfMemoryExceptions. This change adds a check on the container size before we start reading any elements.

### CODE REVIEW

File: core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  // helper method to set container size correctly in the supplied byte array
+  protected void setContainerSize(byte[] buf, int n) {
+    byte[] b = ByteBuffer.allocate(4).putInt(n).array();
+    for (int i = 0; i < 4; i++) {
+      buf[i] = b[i];
+    }
+  }
+
+  @Override
+  public int readAll(byte[] buf, int off, int len) throws TTransportException {
```
Comment (isnotinvain) [inline core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java]: would it make sense to use a mocking library like easymock here?

**Comment by isnotinvain** [comment]
so this throws when the size is reported to be bigger than the remaining bytes? Is that the remaining byte in the record, or in the file?

File: core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java (hunk 1 above)
Comment (rubanm) [inline core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java]: Updated to use easymock.

**Comment by rubanm** [comment]
It's the remaining bytes in the record, which is the size of the byte array supplied to the fromBytes method in the deserializer.

File: core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @Test
+  public void testCheckContainerSizeValid() throws TException {
+    // any non-negative value is considered valid when checkReadLength is not enabled
+    TTransport transport;
+    ThriftBinaryProtocol protocol;
+
+    transport = getMockTransport(3);
+    replay(transport);
```
Comment (isnotinvain) [inline core/src/test/java/com/twitter/elephantbird/thrift/TestThriftBinaryProtocol.java]: once you're done with the mock, call verify(mock) on it as well

**Comment by isnotinvain** [comment]
One minor comment, but LGTM +1

**Comment by isnotinvain** [comment]
I think we could also add a parameter for MAX_COLLECTION_SIZE so that users can also just set a hard max, but that can be done separately.

**Comment by isnotinvain** [comment]
+1

**Comment by rangadi** [comment]
Is this really useful? container size is not the size of the bytes..

**Comment by rubanm** [comment]
This was added due to a OOM issue we ran into internally with invalid container size. This check is used as a heuristic -- if the container size is greater than the number of remaining bytes, then it's clearly corrupt.

**Comment by rangadi** [comment]
Yep! thanks. containers have the same problem as Strings.. thrift first allocates large container first.

**Pull request merged**